### PR TITLE
Handle `Z` correctly in IXDTF

### LINKED ISSUE #5533: `CustomTimeZone` is too unstructured
```rust
// Current
pub trait TimeZoneInput {
    fn offset(&self) -> Option<UtcOffset>;
    fn time_zone_id(&self) -> Option<TimeZoneBcp47Id>;
    fn metazone_id(&self) -> Option<MetazoneId>;
    fn zone_variant(&self) -> Option<ZoneVariant>;
}
```

The `TimeZoneInput` trait combines too much information, and using it as the time zone in `ZonedDateTimeFormatter` yields an unusable API.

### Obervation 1

Datetime libraries deal only with IANA names and offsets:

* `jiff::Timezone`: `iana()`, `to_offset(Timestamp)`
* `chrono_tz::Tz`: `name()`, `offset_from_utc_datetime(DateTime<Utc>)` (chrono doesn't have a timestamp type)
* IXDTF: `-06:00`, `+5:00[America/Vancouver]`, `[America/Vancouver]`
 
The concept of metazones, and zone variants is a CLDR-internal concept.[^2]

[^2]: TZDB has a DST flag, but this is not generally useful, as Ireland for example sets it in the winter. It just means "this is the *other* offset for this location", and is more of a zone-calculation implementation detail.

### Observation 2

The BCP47 ID is superior to the IANA ID. It is fixed width and alias-free. Using it internally is the right choice.

### Observation 3

The formatters currently use `ExtractedTimeZoneInput` and `CustomTimeZone` interchangeably, as those types have the same fields. This should be cleaned up (it should be `ExtractedTimeZoneInput` where it is `ExtractedDate/TimeInput` for the other fields). For the sake of this discussion, please assume that this cleanup happened.

## Proposal

I propose changing the `TimeZoneInput` trait to replace `time_zone_id` by `location`, with a different signature, and remove the `metazone_id` and `zone_variant` methods.

```rust
// Proposed
pub trait TimeZoneInput {
    fn offset(&self) -> Option<UtcOffset>;
    fn location(&self, lookup: Option<&TimezoneIdMapper>) -> Option<TimeZoneBcp47Id>;
}
```

Passing an optional `&TimeZoneIdMapper`, which is managed by the formatter (and enabled with a method like `with_iana_support[_unstable]`) to the `location` methods removes the need for trait implementers to manage their own `TimeZoneIdMapper`. Currently implementing `TimeZoneInput` on, for example, `jiff::Timezone` is not possible; it would require a newtype that wraps `(jiff::Timezone, &TimezoneIdMapper)`, and a way to construct these from `jiff:Timezone`s, some kind of `TimeZoneIdMapperHolder`. Yuck. This is a problem for most implementers, see observation 1.

The second change is the removal of `metazone_id` and `zone_variant`. These fields need a `MetazoneCalculator`, and, more importantly, a `DateTime` to be computed. If they only needed the metazone calculator, we could follow the same approach as with the timezone mapper, but because they need to be computed at a datetime, this is very display-oriented data and not really part of the "identity" of a timezone. In addition to this, the concept of a metazone is not universal, it was invented by CLDR to deduplicate non-location display names. Exposing this directly to users/trait implementers requires more CLDR knowledge than necessary.

Instead, the formatter should load a `MetazoneCalculator` whenever it loads a non-location style. The metazone calculation can then be done just for the lookup in the non-location display names.

This new trait is implementable by time zone abstractions in the ecosystem (`jiff::Timezone`, `chrono_tz::Tz`).

We could then remove the `metazone` and `zone_variant` fields from `icu::timezone::CustomTimeZone` as well. They are already not populated when parsing an IXDTF string, which I consider the main use case of `CustomTimeZone`, and need to be manually computed from a `MetazoneCalculator` *and a datetime*. I also support this change.

### LINKED ISSUE #5653: Consider not using `Etc/Unknown` time zone
The `Etc/Unknown` timezone is not a TZDB timezone. It was invented by CLDR to signal the absence of time zone information, and this is causing confusion when surfaced to users: https://www.google.com/search?q=etc%2Funknown, https://github.com/tc39/proposal-temporal/issues/2510. 

Using `Etc/Unknown` means that every API that handles a time zone input has to special-case this input, which is easy to miss. A more solid API would use `Option<TimeZoneBcp47Id>` with `None` instead of `Etc/Unknown`.

Our `TimeZoneMapper` already returns `Option<TimeZoneBcp47Id>`, and our formatter input already uses `Option<TimeZoneBcp47Id>` so we effectively have *two ways* of representing unknown time zones: `None` and `Some("unk")`. During formatting, `None` will fall back to the offset (good), and then to `{GMT+?}` (good), whereas `Some("unk")` will format as `Unknown City Daylight Time` or `heure: ville inconnue` under location formats *even when there is an offset that could be formatted*.[^1]

My concrete proposal is:
* Remove the `Etc/Unknown` key (and its alias, `Factory`) from the IANA->BCP47 mapping. 
  * Asking for the BCP47 ID for the `Etc/Unknown` IANA ID will fail, because `Etc/Unknown` is not a valid IANA ID
  * Asking for the BCP47 ID for the `Factory` IANA ID will fail because we say so (and document it)
* Remove the `unk` key from the BCP47->IANA mapping
  * It's incorrect to have it in there, because `Etc/Unknown` is not a canonical IANA ID
* Make the location formats fail if there's no exemplar city for a time zone, so that the next format can be tried. The location formats currently produce very awkward strings for the `Etc/Unknown` time zone, even when an offset could be formatted.

[^1]: I understand [UTS-35](https://unicode.org/reports/tr35/tr35-dates.html#Contents:~:text=falling%20back%20to%20the%20localized%20exemplar%20city%20for%20the%20unknown%20zone%20(Etc/Unknown)%2C%20for%20example%20%22Unknown%20City%22%20for%20English) wants it like this, but I don't see why.

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #5533, fixes #5653

### CODE REVIEW

File: components/timezone/src/ixdtf.rs [hunk 1]
```diff
@@ -41,6 +41,23 @@ pub enum ParseError {
     MismatchedTimeZoneFields,
     /// An unknown calendar was provided.
     UnknownCalendar,
+    /// A timezone calculation is required to interpret this string, which is not supported.
```
Comment (Manishearth) [inline components/timezone/src/ixdtf.rs:44]: "which is not supported" feels confusing to me, is the specific time zone not supported, is calculating timezones not supported by ICU4X, is calculating timeones not supported by ICU4X in this specific API, or ?

Seems like the last one, if it is the last one then we should suggest what folks should do instead.

**Review by Manishearth — APPROVED** [review]
code seems correct, want some clarity in docs, ok as followup

File: components/timezone/src/ixdtf.rs (hunk 1 above)
Comment (robertbastian) [reply components/timezone/src/ixdtf.rs:44]: There's an example suggesting what to do instead.

Comment (Manishearth) [reply components/timezone/src/ixdtf.rs:44]: Not really; it suggests changing the string. Someone seeing this error may not have control over the string, we should have solid advice for them when they see this error. Perhaps it is "we don't support `Z` plus a specific timezone"

It feels strange for a timezone calculation library to complain that something requires calculation.

Comment (robertbastian) [reply components/timezone/src/ixdtf.rs:44]: Changing the string is the only solution though. We don't support strings of this shape.

> It feels strange for a timezone calculation library to complain that something requires calculation.

### **We are not a timezone calculation library**

**Pull request merged**